# Mobile App to Predict Probability of Winning the Lottery

Many people start playing the lottery for fun, but for some this activity turns into a habit and eventually escalates into additction.

Suppose a medical institute that aims to prevent and treat gambling addictions wants to build a dedicated mobile app to help lottery addicts better estimate their chances of winning. This project aims to create the logical core of the app that calculates probabilities. The app should focus on the __[6/49 lottery]()__ and builds functions that enables users to answer questions like:

* What is the probability of winning the big prize with a single ticket?
* What is the probability of winning the big prize with 40 different tickets (or any other number of tickets)?
* What is the probability of having at least five (or four, or three, or two) winning numbers on a single ticket?

The app should consider historical __[data]()__ coming from the national 6/49 lottery game in Canada.

### Core Functions

Two functions that will be used often within this project are:

* A function that calculates factorials; and
* A function that calculates combinations.

To calcuate factorials:

\begin{equation*}
n! = n * (n-1) * (n-2) * \cdots * 2 * 1
\end{equation*}

In the 6/49 lottery, six numbers are drawn from a set of 49 numbers that range from 1 to 49. The drawing is done *without replacement*.

To find the number of combinations when sampling without replacement and taking only *k* objects from a group of *n objects*:

\begin{equation*}
_{n}C_{k} = {n \choose k} = \frac{n!}{k!(n-k)!}
\end{equation*}

In [1]:
# Factorial equation using recursion
def factorial(n):
    if n == 0:
        return 1
    else:
        return n * factorial(n-1)
    
# Practice using factorial equation
factorial(5)

120

In [2]:
# Combinations equation
def combinations(n, k):
    return factorial(n) / (factorial(k) * factorial(n-k))

### One-ticket Probability

In the 6/49 lottery, a player wins the big prize if the six numbers on their tickets match all the six number drawn.

The code block below calculates the probability of winning the big prize with the various numbers on a single ticket by:

* Writing a function that takes in a python list with the six unique numbers; and
* prints the probability value in a readable way.

In [3]:
def one_ticket_probability(user_array):
    possible_combinations = combinations(49, 6)
    answer = (1 / possible_combinations) * 100
    return print("Your chance of being the big winner with the numbers {} is {:.7f}%.\nIn other words, you have a 1 in {:,} chances to win.".format(user_array, answer, int(possible_combinations)))

# Test the function
one_ticket_probability([1,2,3,4,5,6])
print('\n')
one_ticket_probability([22, 21, 30, 33, 38, 16])

Your chance of being the big winner with the numbers [1, 2, 3, 4, 5, 6] is 0.0000072%.
In other words, you have a 1 in 13,983,816 chances to win.


Your chance of being the big winner with the numbers [22, 21, 30, 33, 38, 16] is 0.0000072%.
In other words, you have a 1 in 13,983,816 chances to win.


### Historical Data Check for Canada Lottery

The app should allow users to compare their ticket against the historical lottery data in Canada and determine whether they would have ever won by now.

In [4]:
import pandas as pd

lottery_canada = pd.read_csv('649.csv')
print(lottery_canada.shape)
lottery_canada.head(5)

(3665, 11)


PRODUCT  DRAW NUMBER  SEQUENCE NUMBER  DRAW DATE  NUMBER DRAWN 1  \
0      649            1                0  6/12/1982               3   
1      649            2                0  6/19/1982               8   
2      649            3                0  6/26/1982               1   
3      649            4                0   7/3/1982               3   
4      649            5                0  7/10/1982               5   

   NUMBER DRAWN 2  NUMBER DRAWN 3  NUMBER DRAWN 4  NUMBER DRAWN 5  \
0              11              12              14              41   
1              33              36              37              39   
2               6              23              24              27   
3               9              10              13              20   
4              14              21              31              34   

   NUMBER DRAWN 6  BONUS NUMBER  
0              43            13  
1              41             9  
2              39            34  
3              43            34  
4              47            45

In [5]:
lottery_canada.tail(5)

PRODUCT  DRAW NUMBER  SEQUENCE NUMBER  DRAW DATE  NUMBER DRAWN 1  \
3660      649         3587                0   6/6/2018              10   
3661      649         3588                0   6/9/2018              19   
3662      649         3589                0  6/13/2018               6   
3663      649         3590                0  6/16/2018               2   
3664      649         3591                0  6/20/2018              14   

      NUMBER DRAWN 2  NUMBER DRAWN 3  NUMBER DRAWN 4  NUMBER DRAWN 5  \
3660              15              23              38              40   
3661              25              31              36              46   
3662              22              24              31              32   
3663              15              21              31              38   
3664              24              31              35              37   

      NUMBER DRAWN 6  BONUS NUMBER  
3660              41            35  
3661              47            26  
3662              34            16  
3663              49             8  
3664              48            17

The data set contains historical data for 3,665 drawings (each row shows data for a single drawing), dating from 1982 to 2018. Each drawing, one can find the six numbers drawn in the following six column:

* `NUMBER DRAWN 1`
* `NUMBER DRAWN 2`
* `NUMBER DRAWN 3`
* `NUMBER DRAWN 4`
* `NUMBER DRAWN 5`
* `NUMBER DRAWN 6`

### Function for Historical Data Check

The code block below provides a function that enables users to compare their ticket against the historical lottery data in Canada and determine whether they would have ever won by now.

The function will print:

* the number of times the user's selected combination occurred in the Canada data set; and
* the probability of winning the big prize in the next drawing with that combination.

In [6]:
# Extract the winning six numbers from the historical data set as Python sets
def extract_numbers(row):
    row = row[4:10]
    row = set(row.values)
    return row

winning_numbers = lottery_canada.apply(extract_numbers, axis=1)
winning_numbers.head()

0    {3, 41, 11, 12, 43, 14}
1    {33, 36, 37, 39, 8, 41}
2     {1, 6, 39, 23, 24, 27}
3     {3, 9, 10, 43, 13, 20}
4    {34, 5, 14, 47, 21, 31}
dtype: object

In [7]:
# A function that checks the historical occurence of the user's selected numbers
def check_historical_occurrence(user_array, historical_numbers):
    user_set = set(user_array)
    check_occurrence = historical_numbers == user_set
    n_occurrences = check_occurrence.sum()
    
    if n_occurrences == 0:
        print("The combination {} has never occured.\nThis doesn't mean it's more likely to occur now.\nYour chances to win the big prize in the next drawing using the combination {} are 0.0000072%.\nIn other words, you have a 1 in 13,983,816 chances to win.".format(user_array, user_array))
        
    else:
        print("The combination {} has occurred {} time(s) in the past.\nYour chances to win the big prize in the next drawing using the combination {} are 0.0000072%.\nIn other words, you have a 1 in 13,983,816 chances to win.".format(user_array, n_occurrences, user_array))
        

In [8]:
test_input = [33, 47, 15, 11, 2, 28]
check_historical_occurrence(test_input, winning_numbers)

The combination [33, 47, 15, 11, 2, 28] has never occured.
This doesn't mean it's more likely to occur now.
Your chances to win the big prize in the next drawing using the combination [33, 47, 15, 11, 2, 28] are 0.0000072%.
In other words, you have a 1 in 13,983,816 chances to win.


In [9]:
test_input = [33, 36, 37, 39, 8, 41]
check_historical_occurrence(test_input, winning_numbers)

The combination [33, 36, 37, 39, 8, 41] has occurred 1 time(s) in the past.
Your chances to win the big prize in the next drawing using the combination [33, 36, 37, 39, 8, 41] are 0.0000072%.
In other words, you have a 1 in 13,983,816 chances to win.


### Multi-ticket Probability

Users should be able to find the probability of winning if they play multiple different tickets.

The following function takes in the number of tickets and prints the probability information depending on the input. 

In [10]:
def multi_ticket_probability(n_tickets):
    n_combinations = combinations(49, 6)
    probability = n_tickets / n_combinations
    percentage_form = probability * 100
    
    if n_tickets == 1:
        print("Your chances to win the big prize with one ticket are {:.6f}%.\nIn other words, you have a 1 in {:,} chances to win.".format(percentage_form, int(n_combinations)))
    else:
        combinations_simplified = round(n_combinations / n_tickets)
        print("Your chances to win the big prize with {:,} different tickets are {:.6f}%.\nIn other words, you have a 1 in {:,} chances to win.".format(n_tickets, percentage_form, combinations_simplified))     

In [11]:
test_inputs = [1, 10, 100, 1000, 10000, 6991908, 13983816]

for test_input in test_inputs:
    multi_ticket_probability(test_input)
    print('------------------------') # output delimiter

Your chances to win the big prize with one ticket are 0.000007%.
In other words, you have a 1 in 13,983,816 chances to win.
------------------------
Your chances to win the big prize with 10 different tickets are 0.000072%.
In other words, you have a 1 in 1,398,382 chances to win.
------------------------
Your chances to win the big prize with 100 different tickets are 0.000715%.
In other words, you have a 1 in 139,838 chances to win.
------------------------
Your chances to win the big prize with 1,000 different tickets are 0.007151%.
In other words, you have a 1 in 13,984 chances to win.
------------------------
Your chances to win the big prize with 10,000 different tickets are 0.071511%.
In other words, you have a 1 in 1,398 chances to win.
------------------------
Your chances to win the big prize with 6,991,908 different tickets are 50.000000%.
In other words, you have a 1 in 2 chances to win.
------------------------
Your chances to win the big prize with 13,983,816 different ti

### Less Winning Numbers - Function

In most 6/49 lotteries, there are smaller prizes if a player's tickets match two, three, four, or five of the six numbers drawn. This means that players might be interested in finding out the probability of having two, three, four, or five winning numbers.

The resulting function takes an integer between 2 and 5 and prints information about the chances of winning depending on the value of that integer.

The function below calculates the probability that a player's ticket matches exactly the given number of winning numbers. The function will not return the probability of having *at least n winning numbers*.

In [12]:
def probability_less_6(n_winning_numbers):
    
    n_combinations_ticket = combinations(6, n_winning_numbers)
    n_combinations_remaining = combinations(43, 6 - n_winning_numbers)
    successful_outcomes = n_combinations_ticket * n_combinations_remaining
    
    n_combinations_total = combinations(49, 6)
    probability = successful_outcomes / n_combinations_total
    
    probability_percentage = probability * 100
    combinations_simplified = round(n_combinations_total/successful_outcomes)
    print("Your chances of having {} winning numbers with this ticket are {:.6f}%.\nIn other words, you have a 1 in {:,} chances to win.".format(n_winning_numbers, probability_percentage, int(combinations_simplified)))

In [13]:
for test_input in [2, 3, 4, 5]:
    probability_less_6(test_input)
    print('---------------------------')

Your chances of having 2 winning numbers with this ticket are 13.237803%.
In other words, you have a 1 in 8 chances to win.
---------------------------
Your chances of having 3 winning numbers with this ticket are 1.765040%.
In other words, you have a 1 in 57 chances to win.
---------------------------
Your chances of having 4 winning numbers with this ticket are 0.096862%.
In other words, you have a 1 in 1,032 chances to win.
---------------------------
Your chances of having 5 winning numbers with this ticket are 0.001845%.
In other words, you have a 1 in 54,201 chances to win.
---------------------------


### Next Steps

For this project, the four main functions were:

* one_ticket_probability() - calculates the probability of winning the big prize with a single ticket
* check_historical_occurrence() - checks whether a certain combination has occurred in the Canada lottery data set
* multi_ticket_probability() - calculates the probability for any number of tickets between 1 and 13,983,816
* probability_less_6() - calculates the probability of having two, three, four, or five winning numbers exactly.

Possible features for a second version of the app include:

* Making the outputs even easier to understand by adding fun analogies (for example, one can find probabilities for strange eents and compare with the chances of winning in lottery; for instance, one can output something along the lines "You are 100 times more likely to be a victim of a shark attack than winning the lottery")
* Combining the one_ticket_probability() and check_historical_occurrence() to output information on probablity and historical occurrence at the same time.
* Create a function similar to probability_less_6() which calculates the probabiltiy of having *at least* two, three, four, or five winning numbers is the sum of these three numbers:
    * The number of successful outcomes for having four winning numbers exactly
    * The number of successful outcomes for havig five winning numbers exactly
    * The number of successful outcomes for having six winning numbers exactly